### PR TITLE
refactor: refactor CoreFailure part-3 (storage)

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

###Issues

the way API, DB, and Crypto errors are handled in logic is a bit chaotic which lead to different implementation of error handling by repositories

###Solutions

there will be a series of PRs to unify error Handling
changes in this PR:
1 - wrapStorageRequest to catch any DB/preferences errors
2 - handling errors in AssetRepository.kt, ClientRepository.kt and ConversationRepository.kt (to avoid big PR)
3 - delete PreferencesResult class and return null on DataNotFound and handle errors with wrapStorageRequest
4 - delete SessionFailure since it's not needed after  wrapStorageRequest

Needs releases with:

- [ ] GitHub link to other pull request

### Testing

WIP

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/ConversationDAO.kt [hunk 1]
```diff
@@ -2,24 +2,25 @@ package com.wire.kalium.persistence.dao
 
 import kotlinx.coroutines.flow.Flow
 
-data class Conversation(
+// TODO: add conversation / user table
```
Comment (vitorhugods) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/ConversationDAO.kt:5]: Doesn't the `Member` table already serve this purpose?

**Review by vitorhugods — CHANGES_REQUESTED** [review]
Very tiny suggestions.

Nice to have the DB operations taken care of. Great work!

Just in order to keep it tidier in the future, I would suggest doing renaming of the Entity classes in a different PR from this. Especially because the failure handling is quite crucial and could be already merged, but renaming/refactoring is prone to a lot of merge conflicts that can hold the PR down.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MemberMapper.kt [hunk 2]
```diff
@@ -5,7 +5,7 @@ import com.wire.kalium.network.api.UserId
 import com.wire.kalium.network.api.conversation.ConversationMember
 import com.wire.kalium.network.api.conversation.ConversationMembersResponse
 import com.wire.kalium.network.api.user.client.SimpleClientResponse
-import com.wire.kalium.persistence.dao.Member as PersistedMember
+import com.wire.kalium.persistence.dao.MemberEntity as PersistedMember
```
[the remark this answers is not in the document]
Comment (MohamadJaara) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MemberMapper.kt:8]: I did revert the Class naming change because of conflicts

**Comment by MohamadJaara** [comment]
This PR for  is ready to review, there is an issue that I am aware of, which is using 
`wrapStorageRequest` on functions that could return null, I will address it in a later PR

File: android/src/main/kotlin/com/wire/kalium/presentation/MainActivity.kt [hunk 3]
```diff
@@ -39,7 +40,14 @@ class MainActivity : ComponentActivity() {
     private fun loginAndFetchConversationList(coreLogic: CoreLogic) = lifecycleScope.launchWhenCreated {
         login(coreLogic.getAuthenticationScope())?.let {
             val session = coreLogic.getSessionScope(it.userId)
-            val conversations = session.conversations.getConversations().first()
+            val conversations = session.conversations.getConversations().let { result ->
+                when(result) {
+                    is GetConversationsUseCase.Result.Failure -> {
+                        throw IllegalStateException(result.storageFailure.rootCause)
```
Comment (gongracr) [inline android/src/main/kotlin/com/wire/kalium/presentation/MainActivity.kt:46]: Not important at all, but wouldn't an `IOException` suit better?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/CoreFailure.kt [hunk 4]
```diff
@@ -30,18 +29,32 @@ sealed class NetworkFailure : CoreFailure() {
...
-            else -> Either.Left(NetworkFailure.ServerMiscommunication(result.kException))
+    return try {
+        when (val result = networkCall()) {
+            is NetworkResponse.Success -> Either.Right(result.value)
+            is NetworkResponse.Error -> when (result.kException) {
+                else -> Either.Left(NetworkFailure.ServerMiscommunication(result.kException))
+            }
         }
+    } catch (e: Exception) {
+        Either.Left(NetworkFailure.ServerMiscommunication(e))
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/CoreFailure.kt:57]: No more crashes due to weird backend responses! 👏

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/asset/AssetRepository.kt [hunk 5]
```diff
@@ -34,28 +34,29 @@ internal class AssetDataSource(
...
+        }.flatMap { assetResponse ->
+            assetMapper.fromUploadedAssetToDaoModel(uploadAssetData, assetResponse).let { assetEntity ->
+                wrapStorageRequest { assetDao.insertAsset(assetEntity) }
+            }.map { assetMapper.fromApiUploadResponseToDomainModel(assetResponse) }
         }
     }
 
     override suspend fun downloadPublicAsset(assetKey: String): Either<CoreFailure, ByteArray> = suspending {
+        // TODO: handle storage error
         val persistedAsset = assetDao.getAssetByKey(assetKey).firstOrNull()
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/asset/AssetRepository.kt:52]: Why aren't we enclosing this also with `wrapStorageRequest { ... }`?

**Review by gongracr — APPROVED** [review]
LGTM, great to see DB exceptions also being handled 🎰 🥇

**Review by vitorhugods — APPROVED** [review]

**Pull request merged**